text/3808-register-tool.md

```diff
@@ -0,0 +1,271 @@
+- `register_attribute_tool` could be named `register_metadata_tool` instead. `register_metadata_tool` makes it clear that tool attributes do not change the meaning of code, but `register_attribute_tool` makes it clearer how the tool is intended to be used.
+
+[crate-attr]: https://github.com/rust-lang/rfcs/pull/3791
+
+# Prior art
+[prior-art]: #prior-art
+
+- [`clang-tidy`], [`pylint`], [`eslint`], and [`review`] (a racket linter) use inline comments. Whether these count as namespacing is debatable; pylint and eslint include their name in the inline comment and clang-tidy does not. `review` allows both `review: ignore` and `lint: ignore`.
+- [Roslyn analyzers] and [gcc] use `#pragma`s. GCC uses `#pragma GCC` and Roslyn uses `#pragma warning`.
+- C and C++ use [vendor attributes], which are very similar to tool attributes, including namespacing. They do not have syntactic ambiguity with items in the type namespace and so do not perform any kind of name resolution; in Rust terms, all vendor attributes are [inert]. Like this RFC, and unlike the current language, C++ mandates that tools do not restrict namespaces they don't recognize.
+- C# uses [attributes][c-sharp attrs], which are [active], not [inert], i.e. they follow normal name resolution rules. Like attribute macros in Rust, they can be defined in user code and are unrelated to external tools.
+- [`Resyntax`] (a racket refactoring tool, like `cargo fix`) does not allow inline configuration; instead it requires you to write an extension to the tool specifying the new behavior in code.
+
+[`clang-tidy`]: https://clang.llvm.org/extra/clang-tidy/#suppressing-undesired-diagnostics
+[`pylint`]: https://pylint.pycqa.org/en/latest/user_guide/messages/message_control.html#block-disables
+[`eslint`]: https://eslint.org/docs/latest/use/configure/rules#using-configuration-comments-1
+[`review`]: https://github.com/Bogdanp/racket-review#usage
+[Roslyn analyzers]: https://johnnyreilly.com/eslint-your-csharp-in-vs-code-with-roslyn-analyzers#deactivate-linting-partially
+[gcc]: https://gcc.gnu.org/onlinedocs/gcc/Diagnostic-Pragmas.html
+[vendor attributes]:  https://en.cppreference.com/w/cpp/language/attributes
+[c-sharp attrs]: https://learn.microsoft.com/en-us/dotnet/standard/attributes/applying-attributes
+[active]: https://rustc-dev-guide.rust-lang.org/attributes.html#non-builtinactive-attributes
+[`Resyntax`]: https://docs.racket-lang.org/resyntax/Refactoring_Rules_and_Suites.html#(part._.Exercising_.Fine_.Control_.Over_.Comments)
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+How does this interact with [proc-macro lints][`proc_macro_lint`]?
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+- We could allow registering tools in Cargo.toml (with a `package.tools` field?). This would avoid duplicating tool registration for each crate in the workspace. This depends on [`--crate-attr`] being stabilized.
```

Comment:
We should also consider this process allowing registering tool-specific `cfg`s (e.g. `cfg(kani)`)

We might want to consider a way for tools to define these entries and for users to pull them in somehow.

A tools table may have enough overlap with [runnable dependencies](https://github.com/rust-lang/cargo/issues/2267) that we may want to develop one with the other in mind.

We need to consider whether this table is a `workspace` table like `[profiles]` or a package table.  For the former, we should auto-copy it to the package on publish like we do `workspace.resolver`.  If the latter, we should consider whether workspace inheritance should be allowed.

Reply:
> We should also consider this process allowing registering tool-specific cfgs (e.g. cfg(kani))

This seems reasonable, but not all tools support cfgs, so doing it for all tools defeats the point of check-cfg. So we would need some way to configure the behavior.

I would be extremely surprised to see someone building an external tool as a dependency. I can mention it I guess but I don't know that it makes sense to go out of our way to support it.

I don't have strong opinions on the other things, but I will mention them in future possibilities.